### PR TITLE
Duplicates backend & integration tests

### PR DESCRIPTION
**Pull request by RubenGeo**
[AB#33530](https://dev.azure.com/redcrossnl/7d17f2a9-3f18-40f1-8c29-98568fbca7e4/_workitems/edit/33530) [AB#33533](https://dev.azure.com/redcrossnl/7d17f2a9-3f18-40f1-8c29-98568fbca7e4/_workitems/edit/33533) [AB#33540](https://dev.azure.com/redcrossnl/7d17f2a9-3f18-40f1-8c29-98568fbca7e4/_workitems/edit/33540) 

This PR adds:
- duplicates status in get registrations endpoint
- get duplicates endpoint to get duplicate information per registration
- integration tests to test them both

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added tests wherever relevant
- [x] I have made sure that all automated checks pass before requesting a review

## Portalicious preview deployment

This PR does not have any preview deployments yet.

### CODE REVIEW

File: services/121-service/src/registration/dto/duplicate.dto.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  @ApiProperty({ example: 1 })
+  public readonly registrationProgramId: number;
+
+  @ApiProperty({ example: 'zeeland' })
+  public readonly scope: string;
+
+  @ApiProperty({ example: ['phoneNumber'] })
+  public readonly attributeNames: string[];
+
+  @ApiProperty({ example: 'true' })
```
Comment (RubenGeo) [inline services/121-service/src/registration/dto/duplicate.dto.ts:21]: I added this to help the frontend with this [frame](https://www.figma.com/design/77QfhqQ2ye51HJGD0mkBtJ/121---Portalicious?node-id=6525-527870&t=F3cbly6kXYQyQsgn-1)

File: services/121-service/src/registration/dto/duplicate.dto.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { ApiProperty } from '@nestjs/swagger';
+
+// TODO: Discuss should this be name DuplicateReponseDto? (as in the guidelines) And that the api returns DuplicateReponseDto[]?
```
Comment (RubenGeo) [inline services/121-service/src/registration/dto/duplicate.dto.ts:3]: An array of response seems weird. It's sort of one response which can have an arrays of duplicates

File: services/121-service/test/registrations/duplicates/get-duplicates.test.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
+import { RegistrationStatusEnum } from '@121-service/src/registration/enum/registration-status.enum';
```
Comment (RubenGeo) [inline services/121-service/test/registrations/duplicates/get-duplicates.test.ts:1]: I did not add any unit tests. The reason for this is 90 % of logic is in the queries. And therefore the scenario's I could think of are all covered in integration tests. Do you see opportunities for unit testing @diderikvw

File: services/121-service/test/registrations/duplicates/get-duplicates.test.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    expect(duplicates[0]).toMatchObject({
+      registrationProgramId: expect.any(Number),
+      scope: '',
+      attributeNames: ['phoneNumber', 'whatsappPhoneNumber'],
+      name: registration2.fullName,
+      isDuplicateAccessibleWithinScope: true,
+    });
+  });
+
+  it(`should not return isDuplicateAccessibleWithinScope is false and no name for a duplicate if the duplicate registration is out of the scope of the user`, async () => {
```
Comment (diderikvw) [inline services/121-service/test/registrations/duplicates/get-duplicates.test.ts:109]:
```suggestion
  it(`should return isDuplicateAccessibleWithinScope is false and no name for a duplicate if the duplicate registration is out of the scope of the user`, async () => {
```

??

File: services/121-service/test/registrations/duplicates/get-registration-duplicate-status.test.ts
Comment (diderikvw) [inline services/121-service/test/registrations/duplicates/get-registration-duplicate-status.test.ts:1]: Also test for 3+ registrations being duplicates and for multiple attributes included in the duplicate check? And if 2 out of 3 are identical?

Also for the other integration test file. 

And cross-program test as you mentioned? Incl. "identical scope names" between programs?

**Review by diderikvw — CHANGES_REQUESTED** [review]
Elegant and powerful solutions!

File: services/121-service/src/migration/1739551473017-duplicate-status.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
...config"."id"="registration"."programFinancialServiceProviderConfigurationId"  LEFT JOIN "121-service"."latest_message" "latestMessage" ON "latestMessage"."registrationId"="registration"."id"  LEFT JOIN "121-service"."twilio_message" "message" ON "message"."id"="latestMessage"."messageId"  LEFT JOIN (SELECT distinct d1."registrationId" FROM "121-service"."registration_attribute_data" "d1" INNER JOIN "121-service"."registration_attribute_data" "d2" ON d1."programRegistrationAttributeId" = d2."programRegistrationAttributeId" AND "d1"."value" = "d2"."value" AND d1."registrationId" != d2."registrationId"  INNER JOIN "121-service"."registration" "registration1" ON d1."registrationId" = "registration1"."id" AND registration1."registrationStatus" != 'declined'  INNER JOIN "121-service"."registration" "registration2" ON d2."registrationId" = "registration2"."id" AND registration2."registrationStatus" != 'declined'  INNER JOIN "121-service"."program_registration_attribute" "pra" ON d1."programRegistrationAttributeId" = "pra"."id" WHERE "d1"."value" != '' AND pra."duplicateCheck" = true) "dup" ON "registration"."id" = dup."registrationId" ORDER BY "registration"."registrationProgramId" ASC`);
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/121-service/src/migration/1739551473017-duplicate-status.ts:17]: Sounds good. It's easier to look at the query in the registration view entity, by the way. This one is autogenerated by typeorm based on that

File: services/121-service/src/registration/registration-view.entity.ts [hunk 6]
```diff
@@ -92,7 +132,9 @@ export class RegistrationViewEntity {
   @ViewColumn()
   public status: RegistrationStatusEnum;
 
-  @ManyToOne((_type) => ProgramEntity, (program) => program.registrations)
+  @ManyToOne((_type) => ProgramEntity, (program) => program.registrations, {
+    onDelete: 'CASCADE',
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/121-service/src/registration/registration-view.entity.ts:136]: This was a mistake. Good catch :)

File: services/121-service/src/registration/registrations.controller.ts [hunk 7]
```diff
@@ -658,6 +659,32 @@ export class RegistrationsController {
...
+  @AuthenticatedUser({ permissions: [PermissionEnum.RegistrationPersonalREAD] })
+  @ApiOperation({
+    summary: '[SCOPED] Gets duplicate registrations for a registration',
+  })
+  @ApiParam({ name: 'programId', required: true, type: 'integer' })
+  @ApiParam({ name: 'referenceId', required: true, type: 'string' })
+  @ApiResponse({
+    status: HttpStatus.OK,
+    description:
+      'Returns duplicate registrations for a registration. NOTE: this endpoint is scoped, depending on program configuration it only returns/modifies data the logged in user has access to.',
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/121-service/src/registration/registrations.controller.ts:672]: The outcome was that we are going to show the duplicates that are out of scope. Just without the name (fullname). But I just copy pasted this description from other endpoints, and I agreed it could use some tweaking

File: services/121-service/src/registration/registrations.service.ts [hunk 8]
```diff
@@ -906,6 +907,49 @@ export class RegistrationsService {
...
+    referenceId: string,
+    programId: number,
+  ): Promise<DuplicateDto[]> {
+    const registration = await this.getRegistrationOrThrow({
+      referenceId,
+      programId,
+    });
+    const duplicates = await this.registrationScopedRepository.getDuplicates({
+      registrationId: registration.id,
+      programId,
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/121-service/src/registration/registrations.service.ts:920]: Program Id is needed here because you could else potentially get duplicates out of your persmission. As in registraition of a different program

File: services/121-service/src/registration/registrations.service.ts [hunk 9]
```diff
@@ -906,6 +907,49 @@ export class RegistrationsService {
...
+      return [];
+    }
+    const referenceIds = duplicates.map((d) => d.referenceId);
+    // Get the full names of the duplicate using the pagination functionality
+    // This is done to avoid because getting fullNames because this is complex logic and in the pagination service this is already implemented
+    // TODO: In the future this can be refactored so this logic lives in the registration repository
+    const registrationViews =
+      await this.registrationsPaginationService.getRegistrationViewsByReferenceIds(
+        {
+          programId,
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/121-service/src/registration/registrations.service.ts:932]: Program Id is needed here because you could else potentially get duplicates out of your persmission. As in registraition of a different program

File: services/121-service/src/registration/repositories/registration-scoped.repository.ts [hunk 10]
```diff
@@ -164,6 +168,99 @@ export class RegistrationScopedRepository extends RegistrationScopedBaseReposito
...
+  public async getDuplicates({
+    registrationId,
+    programId,
+  }: {
+    registrationId: number;
+    programId: number;
+  }): Promise<GetDuplicatesResult[]> {
+    // Added a limit to the number of duplicates to prevent a large amount of duplicates from being returned which could cause performance issues
+    // I think if there are more than this amount of duplicates a user misconfiguration is likely
+    // Should we also add this limit in the frontend somewhere to inform the user or is this not necessary?
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/121-service/src/registration/repositories/registration-scoped.repository.ts:180]: I contacted dom and tys about this and they both stated it's not needed to show anything in the frontend so I am removing this comment

File: services/121-service/src/registration/services/registrations-pagination.service.ts [hunk 11]
```diff
@@ -166,6 +166,9 @@ export class RegistrationsPaginationService {
 
     // PaginateConfig.select and PaginateConfig.relations cannot be used in combi with each other
     // That's why we wrote some manual code to do the selection
+    if (query.path == null) {
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/121-service/src/registration/services/registrations-pagination.service.ts:169]: Good question. I started refactoring something of the paginate query but than I stopped and reverted it and left this thing behing

**Comment by RubenGeo** [comment]
@diderikvw I implmented all feedback. Only think that is still open are the test cases we need to discuss

File: services/121-service/src/registration/registrations.controller.ts [hunk 12]
```diff
@@ -658,6 +659,32 @@ export class RegistrationsController {
...
+  @AuthenticatedUser({ permissions: [PermissionEnum.RegistrationPersonalREAD] })
+  @ApiOperation({
+    summary: '[SCOPED] Gets duplicate registrations for a registration',
+  })
+  @ApiParam({ name: 'programId', required: true, type: 'integer' })
+  @ApiParam({ name: 'referenceId', required: true, type: 'string' })
+  @ApiResponse({
+    status: HttpStatus.OK,
+    description:
+      'Returns duplicate registrations for a registration. NOTE: this endpoint is scoped, it is only possible to request duplicates for a registration that the logged in user has access to. For the duplicates registrations that are returned the "name" property is only vissisble for registrations in the user scope.',
```
Comment (diderikvw) [inline services/121-service/src/registration/registrations.controller.ts:672]:
```suggestion
      'Returns duplicate registrations for a registration. NOTE: this endpoint is scoped, it is only possible to request duplicates for a registration that the logged in user has access to. For the duplicate registrations that are returned the "name" property is only visisble for registrations in the user's scope.',
```

**Review by diderikvw — DISMISSED** [review]

File: k6/tests/findDuplicates100kRegistrations.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+const registrationsModel = new RegistrationsModel();
+const resetPage = new resetModel();
+const loginPage = new loginModel();
+
+const duplicateNumber = 17; // '17' leads to 131k registrations
+const resetScript = 'nlrc-multiple';
+const programId = 2;
+
+// At the time of implementation, the request duration was 12s on the server and 3s on the local machine for 130k registrations and about 8k duplicates
```
Comment (diderikvw) [inline k6/tests/findDuplicates100kRegistrations.js:17]: Was this aligned with PO as acceptance criterium?

File: k6/tests/findDuplicates100kRegistrations.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  );
+  console.log('🚀 ~ getRegistration:', getRegistration.timings.duration);
+
+  const getRegistrationBody = JSON.parse(getRegistration.body);
+  checkAndFail(getRegistrationBody, {
+    'totalItems between 3000 and 10000': (r) =>
+      r.meta.totalItems >= 3000 && r.meta.totalItems <= 10000,
+  });
+
+  // Check if the request duration is less than 5 seconds
```
Comment (diderikvw) [inline k6/tests/findDuplicates100kRegistrations.js:55]: Comment needs updating? (or remove)

File: services/121-service/test/registrations/duplicates/get-duplicates.test.ts
Comment (diderikvw) [inline services/121-service/test/registrations/duplicates/get-duplicates.test.ts:1]: It would be great if I could just see the changes since my last review. However, it seems that these are "lost" since commits were squashed. Was it necessary to do that? Maybe we should agree not to squash commits while code reviewing is in progress. At least not towards the end, when only small changes still need to be reviewed.

File: services/121-service/test/registrations/duplicates/get-duplicates.test.ts [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+    expect(duplicates[1]).toMatchObject({
+      registrationProgramId: expect.any(Number),
+      scope: '',
+      attributeNames: ['phoneNumber'],
+      name: registration3.fullName,
+      isInScope: true,
+    });
+  });
+
+  it(`should not find duplicates on empty values`, async () => {
```
Comment (diderikvw) [inline services/121-service/test/registrations/duplicates/get-duplicates.test.ts:300]: Looks identical to test on line 183.

File: k6/tests/findDuplicates100kRegistrations.js (hunk 13 above)
Comment (RubenGeo) [reply k6/tests/findDuplicates100kRegistrations.js:17]: yes

File: services/121-service/test/registrations/duplicates/get-duplicates.test.ts
Comment (RubenGeo) [reply services/121-service/test/registrations/duplicates/get-duplicates.test.ts:1]: I squashed all commits because I had to rebase the k6 test branch on this branch. And it's a bit easier to do that if you squash your commit first. 

Not sure about a process for this, but we could discuss it if you like. Maybe put it in the parking lot to also involve dom?

File: services/121-service/test/registrations/duplicates/get-duplicates.test.ts (hunk 15 above)
Comment (RubenGeo) [reply services/121-service/test/registrations/duplicates/get-duplicates.test.ts]: Removed :)

File: services/121-service/test/registrations/duplicates/get-duplicates.test.ts
Comment (aberonni) [reply services/121-service/test/registrations/duplicates/get-duplicates.test.ts:1]: @diderikvw @RubenGeo please see https://teams.microsoft.com/l/message/19:ce634fd37f34477d9f98391f70e1f0ef@thread.skype/1723538316567?tenantId=d3ab9790-6ae2-4bd8-aa5e-02864483e7c7&groupId=55f2fa49-7d9b-4c63-b616-b986f3136c3a&parentMessageId=1723538316567&teamName=510%20-%20CVA&channelName=121%20Development&createdTime=1723538316567

Specifically the parts about "marking files as viewed".

To your point, Diderik:

> It would be great if I could just see the changes since my last review. However, it seems that these are "lost" since commits were squashed.

You can never "trust" commits to be truthful in representing what has changed since your last review. For this reason, I have picked up the habit of using the "viewed" feature in GitHub, which solves exactly your problem, but in a much safer way.

This is true as a general good habit because you can't control how other people manage commits on their own branches (especially, for example, when reviewing code made by external contributes, on open source projects etc.)

Comment (elwinschmitz) [reply services/121-service/test/registrations/duplicates/get-duplicates.test.ts:1]: Only chiming in here on this specific comment by @RubenGeo:

> And it's a bit easier to do that if you squash your commit first.

That is not really true. Actually the opposite, I'd think. Because with smaller commits, Git can solve some issues "for you", so you only have to fix the "actual conflicting bits" manually...

On the general topic, I'm not a fan of squashing commits "in general", mainly because I like "small, concise, specifc commits". But that has more to do with "writing/committing code" then with "reviewing code".

For "reviewing code" I also put my trust in GitHub's tools; But for "understanding how we got here", I like to look at the git-history... and that works better with more commits. :)
